File: Javascript/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// Pake Map atau Object (Hash Map)
export var twoSum = function (nums, target) {
  const numMap = new Map(); // Buat map: (key: angka, value: indeks)
  
  for (let i = 0; i < nums.length; i++) { // Loop dari awal sampai akhir
    const num = nums[i];
    const complement = target - num;    // Hitung angka yang dibutuhkan (complement)

    if (numMap.has(complement)) {     // Cek di map, ada gak complement-nya?
      return [numMap.get(complement), i]; // Kalo ada, BALIKIN INDEKSNYA. (numMap.get(complement) adalah indeks si complement)
    }
    
    numMap.set(num, i);  // Kalo gak ada, simpen num dan indeksnya ke map
  }
  
  return []; // Kalo gak ketemu, balikin array kosong (sesuai soal)
};

console.log(twoSum([2, 7, 11, 15], 9));

File: Golang/test/two_sum.go
package test

func TwoSum(nums []int, target int) []int {
	numMap := make(map[int]int) // Buat map: key: angka, value: indeks
	for i, num := range nums { // Loop dari awal sampai akhir
		complement := target - num // Hitung angka yang dibutuhkan (complement)
		if index, found := numMap[complement]; found { // Cek di map, ada gak complement-nya?
			return []int{index, i} // Kalo ada, BALIKIN INDEKSNYA (index adalah indeks si complement, i adalah indeks si num saat ini)
		}
		numMap[num] = i // Kalo gak ada, simpen num dan indeksnya ke map
	}
	return []int{} // Harusnya ketemu, tapi kalo gak ketemu, balikin array kosong (sesuai soal)
}